parity/main.rs

```diff
@@ -169,61 +175,118 @@ fn start() -> Result<String, String> {
 	}
 
 	let cmd = try!(conf.into_command());
-	execute(cmd)
+	execute(cmd, can_restart)
 }
 
+#[cfg(not(feature="stratum"))]
+fn stratum_main(_: &mut HashMap<String, fn()>) {}
+
 #[cfg(feature="stratum")]
-mod stratum_optional {
-	pub fn probably_run() -> bool {
-		// just redirect to the stratum::main()
-		if ::std::env::args().nth(1).map_or(false, |arg| arg == "stratum") {
-			super::stratum::main();
-			true
-		}
-		else { false }
-	}
+fn stratum_main(alt_mains: &mut HashMap<String, fn()>) {
+	alt_mains.insert("stratum".to_owned(), stratum::main);
 }
 
-#[cfg(not(feature="stratum"))]
-mod stratum_optional {
-	pub fn probably_run() -> bool {
-		false
-	}
+#[cfg(not(feature="ipc"))]
+fn sync_main(_: &mut HashMap<String, fn()>) {}
+
+#[cfg(feature="ipc")]
+fn sync_main(alt_mains: &mut HashMap<String, fn()>) {
+	alt_mains.insert("sync".to_owned(), sync::main);
 }
 
-#[cfg(not(feature="ipc"))]
-fn sync_main() -> bool {
-	false
+fn updates_path(name: &str) -> PathBuf {
+	let mut dest = PathBuf::from(default_hypervisor_path());
+	dest.push(name);
+	dest
 }
 
-#[cfg(feature="ipc")]
-fn sync_main() -> bool {
-	// just redirect to the sync::main()
-	if std::env::args().nth(1).map_or(false, |arg| arg == "sync") {
-		sync::main();
-		true
+fn latest_exe_path() -> Option<PathBuf> {
+	File::open(updates_path("latest")).ok()
+		.and_then(|mut f| { let mut exe = String::new(); f.read_to_string(&mut exe).ok().map(|_| updates_path(&exe)) })
+} 
+
+// Starts ~/.parity-updates/parity and returns the code it exits with.
+fn run_parity() -> Option<i32> {
+	use ::std::ffi::OsString;
+	let prefix = vec![OsString::from("--can-restart"), OsString::from("--force-direct")];
+	latest_exe_path().and_then(|exe| process::Command::new(exe)
+		.args(&(env::args_os().chain(prefix.into_iter()).collect::<Vec<_>>()))
+		.status()
+		.map(|es| es.code().unwrap_or(128))
+		.ok()
+	)
+}
+
+const PLEASE_RESTART_EXIT_CODE: i32 = 69;
+
+// Run our version of parity.
+// Returns the exit error code. 
+fn main_direct(can_restart: bool) -> i32 {
+	let mut alt_mains = HashMap::new();
+	sync_main(&mut alt_mains);
+	stratum_main(&mut alt_mains);
+	if let Some(f) = std::env::args().nth(1).and_then(|arg| alt_mains.get(&arg.to_string())) {
+		f();
+		0
 	} else {
-		false
+		match start(can_restart) {
+			Ok(result) => match result {
+				PostExecutionAction::Print(s) => { println!("{}", s); 0 },
+				PostExecutionAction::Restart => PLEASE_RESTART_EXIT_CODE,
+				PostExecutionAction::Quit => 0,
+			},
+			Err(err) => {
+				writeln!(&mut stdio::stderr(), "{}", err).expect("StdErr available; qed");
+				1
+			},
+		}
 	}
 }
 
-fn main() {
-	// Always print backtrace on panic.
-	::std::env::set_var("RUST_BACKTRACE", "1");
-
-	if sync_main() {
-		return;
+fn println_trace_main(s: String) {
+	if env::var("RUST_LOG").ok().and_then(|s| s.find("main=trace")).is_some() {
+		println!("{}", s);
```

Comment:
not sure what you mean here. if it is being used for anything other than information which *should* go onto stdout, then it is being used incorrectly.

if an execution option requires a message to be placed on stderr or in a log, it should do that explicitly, not pass it back to the dispatch context.